trunk/src/app/srs_app_rtmp_conn.cpp

```diff
@@ -755,25 +755,26 @@ srs_error_t SrsRtmpConn::playing(SrsSharedPtr<SrsLiveSource> source)
     set_sock_options();
     
     // Create a consumer of source.
-    SrsLiveConsumer* consumer = NULL;
-    SrsAutoFree(SrsLiveConsumer, consumer);
-    if ((err = source->create_consumer(consumer)) != srs_success) {
+    SrsLiveConsumer* consumer_raw = NULL;
+    if ((err = source->create_consumer(consumer_raw)) != srs_success) {
```

Comment:
https://github.com/ossrs/srs/blob/baf22d01c1a0f22ae5739f94664763d7b92b173f/trunk/src/app/srs_app_source.cpp#L2662-L2682

This is typical case of a potential memory leak maybe introduced by `SrsUniquePtr` compared with `SrsAutoFree`.

`source->create_consumer` can return not null `consumer_raw` value with a not srs_success err.
In this case, the `consumer_raw` will be a memory leak. 

Two ways to overcome this problem:
```suggestion
    if ((err = source->create_consumer(consumer_raw)) != srs_success) {
        srs_freep(consumer_raw); // free consumer_raw anyway;
```
Or use `SrsAutoFree` in this case.

Reply:
If `create_consumer` return an error, then it should make sure the `**consumer` should not be set to any object.

Reply:
the problem is `create_consumer` need to refined, don't create `consumer` before return err.

```suggestion
    srs_error_t SrsLiveSource::create_consumer(SrsLiveConsumer*& consumer) 
 { 
     srs_error_t err = srs_success; 
  
     // There should be one consumer, so reset the timeout. 
     stream_die_at_ = 0; 
     publisher_idle_at_ = 0; 
  
     // for edge, when play edge stream, check the state 
     if (_srs_config->get_vhost_is_edge(req->vhost)) { 
         // notice edge to start for the first client. 
         if ((err = play_edge->on_client_play()) != srs_success) { 
             return srs_error_wrap(err, "play edge"); 
         } 
     } 
     
     consumer = new SrsLiveConsumer(this); 
     consumers.push_back(consumer); 
      
     return err; 
 } 
```

Reply:
Fixed.